File: indian_hrms_compliance/hr/report/task_compliance/task_compliance.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and Contributors
// For license information, please see license.txt

frappe.query_reports["Task Compliance"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
			reqd: 0,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 0,
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
		},
		{
			fieldname: "kra",
			label: __("KRA"),
			fieldtype: "Link",
			options: "KRA",
		},
		{
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "compliance_pct" && data) {
			// Thresholds come from HR Settings — surfaced on each row by the
			// Python side as _warn_pct / _crit_pct. >= warn = green,
			// >= crit = orange, < crit = red. Defaults match HR Settings defaults.
			const warn = data._warn_pct != null ? data._warn_pct : 80;
			const crit = data._crit_pct != null ? data._crit_pct : 50;
			let color = "var(--text-on-red)";
			let bg = "var(--bg-red)";
			if (data.compliance_pct >= warn) {
				color = "var(--text-on-green)";
				bg = "var(--bg-green)";
			} else if (data.compliance_pct >= crit) {
				color = "var(--text-on-orange)";
				bg = "var(--bg-orange)";
			}
			value = `<span style="background-color:${bg};color:${color};padding:2px 8px;border-radius:4px;font-weight:500">${value}</span>`;
		}
		if (column.fieldname === "overdue" && data && data.overdue > 0) {
			value = `<span style="color:#c0392b;font-weight:500">${value}</span>`;
		}
		return value;
	},
};

File: indian_hrms_compliance/hr/report/task_compliance/task_compliance.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

import frappe
from frappe import _
from frappe.utils import add_days, getdate, today


def execute(filters=None):
	filters = filters or {}
	columns = _get_columns()
	data = _get_data(filters)
	return columns, data


def _get_columns():
	return [
		{
			"label": _("Employee"),
			"fieldname": "employee",
			"fieldtype": "Link",
			"options": "Employee",
			"width": 130,
		},
		{
			"label": _("Employee Name"),
			"fieldname": "employee_name",
			"fieldtype": "Data",
			"width": 180,
		},
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": "Company",
			"width": 180,
		},
		{
			"label": _("KRA"),
			"fieldname": "kra",
			"fieldtype": "Link",
			"options": "KRA",
			"width": 130,
		},
		{"label": _("Total"), "fieldname": "total", "fieldtype": "Int", "width": 80},
		{"label": _("Completed"), "fieldname": "completed", "fieldtype": "Int", "width": 100},
		{"label": _("Pending"), "fieldname": "pending", "fieldtype": "Int", "width": 100},
		{"label": _("Overdue"), "fieldname": "overdue", "fieldtype": "Int", "width": 100},
		{
			"label": _("Compliance %"),
			"fieldname": "compliance_pct",
			"fieldtype": "Percent",
			"width": 130,
		},
	]


def _get_data(filters):
	from_date = filters.get("from_date") or add_days(getdate(today()), -30)
	to_date = filters.get("to_date") or getdate(today())

	# Thresholds drive the colour formatter in task_compliance.js — surface them
	# per-row so the JS doesn't need its own frappe.call.
	warn_pct = float(
		frappe.db.get_single_value("HR Settings", "task_compliance_warning_threshold_pct") or 80
	)
	crit_pct = float(
		frappe.db.get_single_value("HR Settings", "task_compliance_critical_threshold_pct") or 50
	)

	conditions = ["goal_type = 'Task Instance'"]
	params = {"from_date": from_date, "to_date": to_date}
	conditions.append("(due_date BETWEEN %(from_date)s AND %(to_date)s)")

	for f, col in (("company", "company"), ("kra", "kra"), ("employee", "employee")):
		if filters.get(f):
			conditions.append(f"{col} = %({f})s")
			params[f] = filters[f]

	where_sql = " AND ".join(conditions)
	today_d = getdate(today())
	params["today_d"] = today_d

	rows = frappe.db.sql(
		f"""
		SELECT
			employee,
			employee_name,
			company,
			kra,
			COUNT(*) AS total,
			SUM(CASE WHEN status = 'Completed' THEN 1 ELSE 0 END) AS completed,
			SUM(CASE WHEN status IN ('Pending', 'In Progress') AND
				(due_date IS NULL OR due_date >= %(today_d)s) THEN 1 ELSE 0 END) AS pending,
			SUM(CASE WHEN status IN ('Pending', 'In Progress') AND
				due_date IS NOT NULL AND due_date < %(today_d)s THEN 1 ELSE 0 END) AS overdue
		FROM `tabGoal`
		WHERE {where_sql}
		GROUP BY employee, kra
		ORDER BY employee, kra
		""",
		params,
		as_dict=True,
	)

	for r in rows:
		r["compliance_pct"] = (
			(r["completed"] / r["total"] * 100.0) if r["total"] else 0.0
		)
		r["_warn_pct"] = warn_pct
		r["_crit_pct"] = crit_pct

	return rows
